[XLA:CPU] Enable general contraction-biasadd-add fusion

kanvi-nervana opened the pull request:
Remove condition blocking contraction-bias-add fusion as this is now supported.

penpornk commented on xla/service/cpu/tests/onednn_matmul_test.cc:
```diff
@@ -1574,6 +1584,103 @@ TEST_F(MatmulTest, BroadcastedAddAfterFusion) {
 
+// Test MM + BiasAdd + Add fusion : F32
+TEST_F(MatmulTest, SimpleTestF32WithBiasAndAddFusion) {
+  const char* matmul_module_str = R"(
+  HloModule matmul.bias.add.test.f32
+  ENTRY matmul.bias.add.test.f32 {
+    arg0.1 = f32[32,32,40,30] parameter(0), parameter_replication={false}
+    arg0.2 = f32[32,32,30,40]parameter(1), parameter_replication={false}
+    dot.7 = f32[32,32,40,40] dot(arg0.1, arg0.2), lhs_batch_dims={0,1}, lhs_contracting_dims={3}, rhs_batch_dims={0,1}, rhs_contracting_dims={2}
+    const.0 = f32[40] constant(15)
+    bcast.1 = f32[32,32,40,40] broadcast(const.0), dimensions={3}
+    add.0 = f32[32,32,40,40] add(dot.7,bcast.1)
+    const.1 = f32[32,32,40,40] constant(0.65)
+    add.1 = f32[32,32,40,40] add(add.0, const.1)
+    tuple.12 = (f32[32,32,40,40]) tuple(add.1)
+    ROOT get-tuple-element.13 = f32[32,32,40,40] get-tuple-element(tuple.12), index=0
+  })";
```
This module is used 3 times for F32, B16, and F16. Could you please create a helper function that calls `absl::StrReplaceAll` based on the input dtype? E.g., for bf16 we could call `CreateMatmulBiasAddAndAddModuleText("f32", "bf16")` and in the function we have a template string that supports at most 2 data types:
```
    ...
    arg0.1 = ${dtype1}[32,32,40,30] parameter(0), parameter_replication={false}
    convert.0 = ${dtype2}[32,32,40,30] convert(arg0.1)
    ...
```
etc.

In the long run, I think we should set up parameterized tests, but we can do that later. I can create an example in this file after this PR is in.

penpornk commented on xla/service/cpu/tests/onednn_matmul_test.cc:
```diff
@@ -1574,6 +1584,103 @@ TEST_F(MatmulTest, BroadcastedAddAfterFusion) {
   )");
 }
 
+// Test MM + BiasAdd + Add fusion : F32
```
Nit: MM -> MatMul
```suggestion
// Test MatMul + BiasAdd + Add fusion : F32
```
Same for other places.

penpornk reviewed (changes requested):
I'm very sorry for the delay. Could you please help resolve conflicts?

kanvi-nervana commented:
> I'm very sorry for the delay. Could you please help resolve conflicts?

The change in the test file hasn't made it to the internal repo yet, once it does I will resolve it. Thanks!

kanvi-nervana commented:
@penpornk I resolved the conflicts. Thanks!

dimitar-asenov commented:
@penpornk Could you please review this PR again?

derdrdirk commented:
I am currently reviewing the PR and trying to fix a failing unit test.

The test can be found in google3/third_party/tensorflow/compiler/xla/service/cpu/tests/onednn_convolution_test.cc
```
// This test should match BIAS + RESIDUAL ADD when the residual add fusion is
// re-enabled.
TEST_P(ConvolutionTest, Conv2DWithBiasAndBinaryAddTest) {
  const absl::string_view outline = R"(
  HloModule convolution.add.test

  ENTRY convolution.add.test {
    arg0.1 = $dtype[1,22,22,1] parameter(0)
    arg0.2 = $dtype[8,8,1,10] parameter(1)
    convolution.0 = $dtype[1,11,11,10] convolution(arg0.1, arg0.2),
          window={size=8x8 stride=2x2 pad=3_3x3_3}, dim_labels=b01f_01io->b01f
    const.0 = $dtype[10] constant(15)
    bcast.1 = $dtype[1,11,11,10] broadcast(const.0), dimensions={3}
    add.0 = $dtype[1,11,11,10] add(convolution.0, bcast.1)
    const.1 = $dtype[1,11,11,10] constant({...})
    ROOT add.1 = $dtype[1,11,11,10] add(add.0, const.1)
  })";

  RunCompareAndMatchOptimizedHlo(outline, {"BIAS"});
}
```

I following the test comment I added "BINARY_ADD", but this fixes only the $dtype=F32 case not the $dtype=BF16 case, which errors as:
```
[ RUN      ] OneDnnConvolutionTestSuite/ConvolutionTest.Conv2DWithBiasAndBinaryAddTest/F32
2025-04-11 09:29:02.402954: E xla/hlo/testlib/filecheck.cc:89] Tried to execute FileCheck at /b/f/w/bazel-out/k8-opt/bin/xla/service/cpu/tests/onednn_convolution_test.runfiles/xla/external/llvm-project/llvm/FileCheck
2025-04-11 09:29:02.403005: E xla/hlo/testlib/filecheck.cc:95] FileCheck input was:
HloModule convolution.add.test, entry_computation_layout={(f32[1,22,22,1]{3,2,1,0}, f32[8,8,1,10]{3,2,1,0})->f32[1,11,11,10]{3,2,1,0}}

ENTRY %convolution.add.test (arg0.1: f32[1,22,22,1], arg0.2: f32[8,8,1,10]) -> f32[1,11,11,10] {
  %arg0.1 = f32[1,22,22,1]{3,2,1,0} parameter(0)
  %arg0.2 = f32[8,8,1,10]{3,2,1,0} parameter(1)
  %const.0 = f32[10]{0} constant({15, 0, 0, 0, 0, 0, 0, 0, 0, 0})
  %const.1 = f32[1,11,11,10]{3,2,1,0} constant({...})
  %copy.1 = f32[1,11,11,10]{3,2,1,0} copy(%const.1)
  %custom-call.2.clone = (f32[1,11,11,10]{3,2,1,0}, u8[241688]{0}) custom-call(%arg0.1, %arg0.2, %const.0, %copy.1), custom_call_target="__onednn$convolution", output_to_operand_aliasing={{0}: (3, {})}, backend_config={"outer_dimension_partitions":[],"onednn_conv_config":{"dims":"4","input":{"dims":"0","data":{"batch_dim":"0","feature_dim":"3","spatial_dims":["2","3"]}},"kernel":{"dims":"0","filter":{"input_feature_dim":"2","output_feature_dim":"3","spatial_dims":["1","2"],"shape":[]}},"output":{"dims":"0","data":{"batch_dim":"0","feature_dim":"3","spatial_dims":["2","3"]}},"window":{"size":[],"pad_left":["4","4"],"pad_right":["4","4"],"strides":["3","3"],"window_dilations":["2","2"]},"fusions":{"ops":["BIAS","SUM"],"alpha_typecast":0},"feature_groups":"1","optimization_config":{"weights_prepacked":false,"user_scratchpad":true,"bias_broadcast":true}}}
  ROOT %get-tuple-element = f32[1,11,11,10]{3,2,1,0} get-tuple-element(%custom-call.2.clone), index=0
}


2025-04-11 09:29:02.403015: E xla/hlo/testlib/filecheck.cc:96] FileCheck stderr:
/b/f/w/_tmp/d16948e503c2c3614888d7f4ba369e19/tempfile-localhost-c5bb632a989e0c44-15-6327d55398364:7:15: error: CHECK-DAG: expected string not found in input
 ; CHECK-DAG: "ops":["BIAS","BINARY_ADD"]
 ```

Can you fix the test or confirm that for BF16 we want to check for ["BIAS","SUM"]?

derdrdirk commented:
Friendly ping :)

kanvi-nervana commented:
> Friendly ping :)

Sorry, for the late reply. I rebased the branch and added a commit for the fix. Basically, the optimized HLO must match SUM for precisions supporting elementwise add operations.

derdrdirk commented:
Looks like there are some branch conflicts left that need to be addressed.

kanvi-nervana commented:
> Looks like there are some branch conflicts left that need to be addressed.

Done, the conflict did not show up internally earlier.

kanvi-nervana commented:
@derdrdirk Gentle reminder :)
I was wondering if there is any update on this PR?

derdrdirk commented:
I just approved the Git workflow. Seems like there are some formatting checks failing.

kanvi-nervana commented:
> I just approved the Git workflow. Seems like there are some formatting checks failing.

Sorry, I missed a conflict.
